File: src/main/examples/scala/oscar/ElementTest.scala
package oscar

import checker.CPChecker._
import checker._
import checker.filterings.HybridFiltering
import oscar.algo.Inconsistency
import oscar.cp.{CPIntVar, CPSolver}
import oscar.cp.constraints.{ElementCst, ElementVarAC}
import oscar.cp.core.CPPropagStrength

/*
 * Testing OscaR's element constraint x[i]==v with a fixed array x
 */
object ElementTest extends App {
  // set the fixed array x
  val x: Array[Int] = Array(1, 5, 4, 2, 3)

  // OscaR's filtering to be tested
  val testedFiltering: Filter = new Filter {
    override def filter(variables: Array[Set[Int]]): Array[Set[Int]] = elementACFiltering(variables)
  }

  // setting the number of tests to 124
  testArguments.setNbTests(124)
  //Then we set x with a size of 7
  testArguments.setNVar(2)
  //add variable i in generator
  testArguments.setRange(0, (0, 5))
  testArguments.setDensity(0, 0.5)
  //add variable v in generator
  testArguments.setRange(1, (1, 6))
  testArguments.setDensity(1, 0.5)
  // setting the seed
  testArguments.setSeed(123456)

  // creation of the trusted algorithm. Since OscaR's filtering reaches range consistency
  // for the i variable and bound(D) consistency for the v variable, we use the
  // hybridFiltering here with (4,3) which means that the first variable must reach
  // range consistency and the second one bound(D) consistency.
  // (to recall, 1 represents AC, 2 for BC(Z), 3 for BC(D) and 4 for RC).
  val trustedFiltering: Filter = new HybridFiltering(Array(4, 3), elementChecker)

  // checking that the filtered domains returned by both algorithms are the same over some
  // random test instances
  check(trustedFiltering, testedFiltering)

  /*
   * This function apply the ElementVarAC constraint of OscaR on the variables
   * passed in argument in this format: vars = [x1,x2,...xn, i,v]
   * It then return those variables filtered
   */
  private def elementACFiltering(vars: Array[Set[Int]]): Array[Set[Int]] = {
    implicit val testSolver: CPSolver = CPSolver(CPPropagStrength.Strong)
    val i = CPIntVar(vars(0))
    val v = CPIntVar(vars(1))
    val ad = new ElementCst(x, i, v)
    try {
      ad.setup(CPPropagStrength.Strong)
      //for(i <- 1 to 10)
      //ad.propagate()
      testSolver.post(ad)
    } catch {
      case _: Inconsistency => throw new NoSolutionException
    }
    Array(i.toArray.toSet, v.toArray.toSet)
  }

  /*
   * This function return true if the solution passed in argument
   * respect the element constraint with its last two element being
   * the variables i and v (solution = [x1,x2,..xn, i,v]
   */
  private def elementChecker: Array[Int] => Boolean = {
    solution => {
      val i = solution(0)
      val v = solution(1)
      i < 5 && i >= 0 && v == x(i)
    }
  }
}

File: src/main/examples/java/choco/ElementTest.java
package choco;

import checker.*;
import checker.Statistics;
import checker.filterings.ArcFiltering;
import org.chocosolver.solver.Model;
import org.chocosolver.solver.Solver;
import org.chocosolver.solver.constraints.ConstraintsName;
import org.chocosolver.solver.constraints.binary.element.PropElement;
import org.chocosolver.solver.variables.IntVar;

import java.util.HashSet;
import java.util.Set;
import java.util.function.Function;

/*
 * Testing the choco's filtering algorithm for the constraint x[i] == v for a fixed x.
 */
public class ElementTest {

    // The array x considered for the element constraint x[i] == v with a fixed array x
    private static int[] x = {5, 1, 2, 4, 7, 9};

    public static void main(String[] args) {

        // choco's filtering to be tested for the constraint x[i]==v
        class testedFilter extends JFilter {
            public Set<Integer>[] filterJava(Set<Integer>[] variables) {
                Model model = new Model("Testing choco's element filtering");
                int[] b = variables[0].stream().mapToInt(Number::intValue).toArray();
                IntVar in = model.intVar(b);
                b = variables[1].stream().mapToInt(Number::intValue).toArray();
                IntVar v = model.intVar(b);
                org.chocosolver.solver.constraints.Constraint cstr = new org.chocosolver.solver.constraints.Constraint(ConstraintsName.ELEMENT,
                        new PropElement(v, x, in, 0));
                model.post(cstr);
                Solver solver = model.getSolver();
                try {
                    solver.propagate();
                } catch (Exception e) {
                    throw new NoSolutionException("");
                }
                IntVar[] finalVars = {in, v};
                return transform(finalVars);
            }
        }
        // sets the test parameters
        TestArgs parameters = new TestArgs();
        parameters.setNVar(2);
        parameters.setDensity(0, 0.5);
        parameters.setRange(0, 0, 6);
        parameters.setDensity(1, 0.5);
        parameters.setRange(1, 0, 10);
        Statistics stats = new Statistics("");

        // the trusted algorithm to be compared with the choco ones
        ArcFiltering trusted = new ArcFiltering(elementChecker());

        // checking the equality of both algorithms over some random domains
        CPChecker.check(trusted, new testedFilter(), parameters, stats);
    }

    private static Function<Integer[], Boolean> elementChecker() {
        return solution -> {
            int i = solution[0];
            int v = solution[1];
            if (i < 0 || i >= x.length) return false;
            else return x[i] == v;
        };
    }


    /*
     * returns the domains in the type Set<Integer> from the choco
     * domains' type IntVar
     */
    private static Set<Integer>[] transform(IntVar[] input) {
        Set<Integer>[] result = new Set[input.length];
        for (int i = 0; i < input.length; i++) {
            result[i] = new HashSet<Integer>();
        }
        for (int i = 0; i < input.length; i++) {
            int elem = input[i].getLB();
            int ub = input[i].getUB();
            while (elem != ub) {
                result[i].add(elem);
                elem = input[i].nextValue(elem);
            }
            result[i].add(ub);
        }
        return result;
    }
}
